### PR TITLE
Coccinelle: Add support for coccinelle infrastructure

### PR DESCRIPTION
**Pull request by himanshujha199640**

### CODE REVIEW

File: Makefile
Comment (dbkinder) [inline Makefile]: I think the intent of this Makefile is only as a convenience for the CI process generating the documentation.   I'll let the dev team speak up about this.

File: doc/scripts/coccinelle.rst
Comment (dbkinder) [inline doc/scripts/coccinelle.rst]: Shouldn't this be talking about the Zephyr project's code base and not the Linux kernel?

File: doc/scripts/coccinelle.rst
Comment (dbkinder) [inline doc/scripts/coccinelle.rst]: There is no such folder in the Zephyr code tree.

File: doc/scripts/coccinelle.rst
Comment (dbkinder) [inline doc/scripts/coccinelle.rst]: Check out the Zephyr doc writing conventions at http://docs.zephyrproject.org/contribute/doc-guidelines.html.  For example, documents start with a label that allows cross-document linking:

    .. _coccinelle:

And our heading hierarchy is #, *, =, -

That aside, I'm not sure this is the right level of documentation for the Zephyr project.  It makes assumptions that the reader knows about Coccinelle and doesn't explain why a Zephyr developer should care or want to use this (external) tool?  What does it tell them about the Zephyr code it's scanning?

The examples and explanations are all specific to the Linux kernel and not the Zephyr project.  (If this is just a copy of Coccinelle documentation, then it would be better to have some Zephyr-specific information here, and point to the details over in your documentation.)  Examples here should be Zephyr-specific and help explain the value of this tool to the Zephyr developer.

Comment (dbkinder) [reply doc/scripts/coccinelle.rst]: The ``doc/scripts`` folder is where we have scripts used to prepare documentation. Your documentation would be better placed in the ``doc/application`` folder and linked into the "Developer Guides" TOC (the index.rst file that's in the doc/application folder.

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: Makefile
Comment (ulfalizer) [inline Makefile]: Kbuild is the name for the Makefiles in the Linux kernel, so this comment might look confusing.

**Review by ulfalizer — CHANGES_REQUESTED** [review]

**Comment by himanshujha199640** [comment]
Cc'ing Julia Lawall,

On Thu, Aug 30, 2018 at 04:44:25PM -0700, David Kinder wrote:
> dbkinder requested changes on this pull request.
> 
> 
> 
> > @@ -15,3 +15,17 @@ export Q
>  # ---------------------------------------------------------------------------
>  htmldocs:
>  	$(Q)$(MAKE) -C doc htmldocs
> +
> 
> I think the intent of this Makefile is only as a convenience for the CI process generating the documentation.   I'll let the dev team speak up about this.

The future of Coccinelle would be integration into the Zephyr CI as we
had proposed in TSC meeting.

The CI will report the warnings/errors found in the source to Julia and
if the warning/errors are not false positives, then she may propose the
patch. And if it is a false postive then we shall improve the
"Confidence" of the cocci rule by re-working the SmPL script.

This similar approach is followed in the kernel community and
0-DAY kernel test infrastructure(Open Source Technology Center @ Intel Corporation)
https://01.org/lkp/documentation/0-day-test-service
also runs Coccinelle and sends report to Julia. If the report is valid,
she added her SoB while sending the patch to relevant maintainers.
https://lore.kernel.org/lkml/alpine.DEB.2.21.1808311158220.2328@hadrien/

So, whenever someone would open a PR for Zephyr, we could run all the
SmPL rules on the PR and report the committer with warnings/errors.

> > +https://github.com/coccinelle/coccinelle/blob/master/install.txt
> +
> +Supplemental documentation
> +---------------------------
> +
> +For supplemental documentation refer to the wiki:
> +
> +https://bottest.wiki.kernel.org/coccicheck
> +
> +The wiki documentation always refers to the linux-next version of the script.
> +
> +For Semantic Patch Language(SmPL) grammar documentation refer to:
> +
> +http://coccinelle.lip6.fr/documentation.php
> +
> +Using Coccinelle on the Linux kernel
> 
> Shouldn't this be talking about the Zephyr project's code base and not the Linux kernel?

Yes, I will change all those instances to Zephyr specific ones
in the next version.

> > +	make coccicheck COCCI=<my_SP.cocci> MODE=patch
> +
> +or::
> +
> +	make coccicheck COCCI=<my_SP.cocci> MODE=report
> +
> +
> +Controlling Which Files are Processed by Coccinelle
> +---------------------------------------------------
> +
> +By default the entire kernel source tree is checked.
> +
> +To apply Coccinelle to a specific directory, ``M=`` can be used.
> +For example, to check drivers/net/wireless/ one may write::
> +
> +    make coccicheck M=drivers/net/wireless/
> 
> There is no such folder in the Zephyr code tree.

Yes, I'll take an example from Zephyr source.

> > @@ -0,0 +1,487 @@
> +.. Copyright 2010 Nicolas Palix <npalix@diku.dk>
> 
> Check out the Zephyr doc writing conventions at http://docs.zephyrproject.org/contribute/doc-guidelines.html.  For example, documents start with a label that allows cross-document linking:
> 
>     .. _coccinelle:
> 
> And our heading hierarchy is #, *, =, -

Sure!

> That aside, I'm not sure this is the right level of documentation for the Zephyr project.  It makes assumptions that the reader knows about Coccinelle and doesn't explain why a Zephyr developer should care or want to use this (external) tool?  What does it tell them about the Zephyr code it's scanning?

Well, you certainly wouldn't want me add links to "tutorials".
What part is not clear to you and should be elaborated ?

I'm not sure adding tutorial links is benefiticial since we
have already mentioned link to the SmPL grammar, modes, parallelism,
and stuff.

> The examples and explanations are all specific to the Linux kernel and not the Zephyr project.  (If this is just a copy of Coccinelle documentation, then it would be better to have some Zephyr-specific information here, and point to the details over in your documentation.)  Examples here should be Zephyr-specific and help explain the value of this tool to the Zephyr developer.

I posted once trivial example test here:
https://lists.zephyrproject.org/g/devel/message/5030

We shall soon frame new SmPL rules to capture those cases of double
locking, potential null dereferences etc. This example of "badzero.cocci" 
was just to show the community how it works in various modes and get
some feedback.

> > @@ -0,0 +1,487 @@
> +.. Copyright 2010 Nicolas Palix <npalix@diku.dk>
> 
> The ``doc/scripts`` folder is where we have scripts used to prepare documentation. Your documentation would be better placed in the ``doc/application`` folder and linked into the "Developer Guides" TOC (the index.rst file that's in the doc/application folder. 

Ah, ok.
I will change the location as you wish.


PS: Julia is the main inventor of the Coccinelle infrastructure and
she can answer the queries better than me. I am just an undergraduate
intern working with her and have worked on Coccinelle in the past while
publishing two rules for the mainline and sending some patches using
the same.

So, I would request to Cc all queries/suggestions/rants to Julia too
to get better feedback. 

And does my reply get copied on devel or any other public mailing list ?
It would be good since everyone can look after the conversation thread.


Thanks
-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

**Comment by himanshujha199640** [comment]
On Thu, Aug 30, 2018 at 05:23:51PM -0700, Ulf Magnusson wrote:
> ulfalizer requested changes on this pull request.
> 
> 
> 
> > @@ -15,3 +15,17 @@ export Q
>  # ---------------------------------------------------------------------------
>  htmldocs:
>  	$(Q)$(MAKE) -C doc htmldocs
> +
> +# SHELL used by kbuild
> +CONFIG_SHELL := $(shell if [ -x "$$BASH" ]; then echo $$BASH; \
> +          else if [ -x /bin/bash ]; then echo /bin/bash; \
> +          else echo sh; fi ; fi)
> +
> +ifeq ("$(origin M)", "command line")
> +  KBUILD_EXTMOD := $(M)
> +endif
> +
> +export KBUILD_EXTMOD
> 
> Another Kbuild reference here. Zephyr doesn't support external modules (or modules in general).

This exported variable is not for building external modules
and rather support Coccinelle to only work on some specified
directory using:

Controlling Which Files are Processed by Coccinelle
---------------------------------------------------

By default the entire kernel source tree is checked.

To apply Coccinelle to a specific directory, ``M=`` can be used.
For example, to check drivers/net/wireless/ one may write::

    make coccicheck M=drivers/net/wireless/

In 'scripts/coccicheck' :

if [ "$KBUILD_EXTMOD" = "" ] ; then
        OPTIONS="--dir $ZEPHYR_BASE $COCCIINCLUDE"
else
        OPTIONS="--dir $KBUILD_EXTMOD $COCCIINCLUDE"
fi

But I agree that the name "KBUILD" is misleading and should be
changed!

Any opinion of new name for the exported variable ?

> Not sure if something like this could be useful for external projects.
> 
> > @@ -15,3 +15,17 @@ export Q
>  # ---------------------------------------------------------------------------
>  htmldocs:
>  	$(Q)$(MAKE) -C doc htmldocs
> +
> 
> Yeah, so far at least.
> 
> Never used Cocinelle myself, but if we want it, it would IMO be fine to have it *nix-only and not write a CMake version (not even sure if it runs on Windows). In that case, the top-level Makefile might be the handiest place to put it.

Coccinelle can run on Windows with limited access to
python or ocaml scripting. But we never have tried in the past
and don't know how well it is supported.

It should certainly work on macOS and Linux ofcourse, so make
seems to me the correct option for now.

But if you wish to make it cmake style then we can do it that
way too. It would just need converting those 5-6 lines of make
to cmake and we are less concerned about it.

The main backbone is the 'coccicheck' shell scripts which
provides the front-end to run coccinelle with various modes,
parallelism, verbose output etc.

Not sure if parmap lib which is used to support parallelism
in coccinelle is supported on Windows.


-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

**Comment by himanshujha199640** [comment]
On Fri, Aug 31, 2018 at 02:23:44AM -0700, Carles Cufí wrote:
> @himanshujha199640 thanks for the PR. This PR makes consistent references in documentation and scripts to UNIX and Linux in particular. Does coccinelle not run on Windows or macOS? If it does, then we should make sure that this is reflected in the code and documentation, since Zephyr runs on all 3 major OS families.

We are doubtless aware that Zephyr runs on all 3 major OS.

Coccinelle support would be minimal for windows but should
run as expected for macOS.

We never had any user of coccinelle on windows and the main
target for coccinelle was the Linux Kernel.

For the documentation, I will make it Zephyr oriented with
examples and references.

Julia(CC'ed) will answer with a more constructive reply about the
Windows/MacOS support.


-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

**Comment by himanshujha199640** [comment]
On Sun, 2 Sep 2018, Himanshu Jha wrote:

> On Fri, Aug 31, 2018 at 02:23:44AM -0700, Carles Cufí wrote:
> > @himanshujha199640 thanks for the PR. This PR makes consistent references in documentation and scripts to UNIX and Linux in particular. Does coccinelle not run on Windows or macOS? If it does, then we should make sure that this is reflected in the code and documentation, since Zephyr runs on all 3 major OS families.
>
> We are doubtless aware that Zephyr runs on all 3 major OS.
>
> Coccinelle support would be minimal for windows but should
> run as expected for macOS.
>
> We never had any user of coccinelle on windows and the main
> target for coccinelle was the Linux Kernel.
>
> For the documentation, I will make it Zephyr oriented with
> examples and references.
>
> Julia(CC'ed) will answer with a more constructive reply about the
> Windows/MacOS support.

Hello,

I know people who use Coccinelle on a Mac, so I think it should be no
problem.  I don't know anything about use on Windows.  Running Linux
within Windows should be a solution if Coccinelle doesn't run directly.

julia

**Comment by carlescufi** [comment]
Hi @himanshujha199640 

>So, I would request to Cc all queries/suggestions/rants to Julia too
to get better feedback.

In that case let's follow the standard practice of adding Julia as a reviewer to this Pull Request. I assume she has a GitHub account? Reviews on GitHub are typically done on the web interface and not via email.

>And does my reply get copied on devel or any other public mailing list ?
It would be good since everyone can look after the conversation thread.

No, the PR itself is the public display of the code changes proposed, that's how it works on GitHub. We don't include review comments in mailing lists.

**Comment by carlescufi** [comment]
@himanshujha199640 

>Coccinelle support would be minimal for windows but should
run as expected for macOS.

Please make that clear. Anytime you use `make` or `bash` or anything that is not CMake or Python you are excluding Windows support. While that is fine if Coccinelle currently doesn't run on Windows well, then please document it accordingly.

>We never had any user of coccinelle on windows and the main
target for coccinelle was the Linux Kernel.

Let's start with support for *NIX and then we can tackle the Windows issue later.

>For the documentation, I will make it Zephyr oriented with
examples and references.

Great, just make sure to mention Windows is not supported, perhaps by adding an item here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/contribute/contribute_guidelines.rst#contribution-tools-and-git-setup

**Comment by himanshujha199640** [comment]
On Mon, Sep 03, 2018 at 03:12:57AM -0700, Carles Cufí wrote:
> Hi @himanshujha199640 
> 
> >So, I would request to Cc all queries/suggestions/rants to Julia too
> to get better feedback.
> 
> In that case let's follow the standard practice of adding Julia as a reviewer to this Pull Request. I assume she has a GitHub account? Reviews on GitHub are typically done on the web interface and not via email.

Yes, please add her to reviewer/maintainers list.
https://github.com/JuliaLawall

> >And does my reply get copied on devel or any other public mailing list ?
> It would be good since everyone can look after the conversation thread.
> 
> No, the PR itself is the public display of the code changes proposed, that's how it works on GitHub. We don't include review comments in mailing lists.

Alright! So, I assume if someone is not a "official reviewer" and wishes
to look at the development in Zephyr, then one has to look at the 'Issues' and
'PR' to determine the direction of development ?

-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

**Comment by himanshujha199640** [comment]
On Mon, Sep 03, 2018 at 10:18:43AM +0000, Carles Cufí wrote:
> @himanshujha199640 
> 
> >Coccinelle can run on Windows with limited access to
> python or ocaml scripting. But we never have tried in the past
> and don't know how well it is supported.
> 
> Let's avoid it for now as mentioned above.
> 
> >It should certainly work on macOS and Linux ofcourse, so make
> seems to me the correct option for now.
> 
> Agreed.
> 
> >But if you wish to make it cmake style then we can do it that
> way too. It would just need converting those 5-6 lines of make
> to cmake and we are less concerned about it.
> 
> Yes, I would rather you used CMake instead of `make` actually. This is more consistent with the rest of the RTOS, regardless of whether Coccinelle runs on Windows or not.

OK. I will try to use CMake istead but have never used in the
past, so I need to lookout for some tutorials/documentation
to reconstruct those 5-6 lines of Makefile. Any references ?

Is there any rule/guide to follow while writing CMake file for
Zephyr ? And this CMake shall be in the top-level since we enable
partial parsing of specific directory using M=./path/to/directory
i.e., suppose we only want to run Coccinelle on drivers/serial, in
such case simply run:

$ make coccicheck M=./drivers/serial/


-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

**Review by carlescufi — DISMISSED** [review]
NACKing based on comments made

**Comment by SebastianBoe** [comment]
> that's a good question. We need a CMake expert for that, so paging @SebastianBoe

Hi, after looking into the make-integration it seems that there is no reason to integrate with the build system at all. So I would recommend organizing this as a bash script instead, like e.g. sanitycheck.

So the user would do ~/zephyr/scripts/coccicheck M=path/to/subdir

If there are no plans for making this portable, and it does not use any information from the build system I don't see any benefits of Make over CMake, but several downsides; e.g. an inconvenient CLI.

As an aside, I was not able to make it run using coccinelle 1.0.4.deb-2. It prints a bunch of compare pointer with 0 warnings and then errors out with "coccicheck failed".

**Comment by JuliaLawall** [comment]
On Tue, 4 Sep 2018, Sebastian Bøe wrote:

>
>       that's a good question. We need a CMake expert for that, so
>       paging @SebastianBoe
>
> Hi, after looking into the make-integration it seems that there is no reason
> to integrate with the build system at all. So I would recommend organizing
> this as a bash script instead, like e.g. sanitycheck.
>
> So the user would do ~/zephyr/scripts/coccicheck M=path/to/subdir
>
> If there are no plans for making this portable, and it does not use any
> information from the build system I don't see any benefits of Make over
> CMake, but several downsides; e.g. an inconvenient CLI.
>
> As an aside, I was not able to make it run using coccinelle 1.0.4.deb-2. It
> prints a bunch of compare pointer with 0 warnings and then errors out with
> "coccicheck failed".

1.0.4 is quite old.  I see that the experimental version of Debian has
1.0.6.  We are in the midst of releasing 1.0.7.

julia

**Comment by himanshujha199640** [comment]
On Tue, Sep 04, 2018 at 04:19:20AM -0700, Sebastian Bøe wrote:
> > that's a good question. We need a CMake expert for that, so paging @SebastianBoe
> 
> Hi, after looking into the make-integration it seems that there is no reason to integrate with the build system at all. So I would recommend organizing this as a bash script instead, like e.g. sanitycheck.
> 
> So the user would do ~/zephyr/scripts/coccicheck M=path/to/subdir

Hmm. But we don't have such facility in shell-scripts to assign
variables via commandline "M=./path/" or "V=1" or "MODE=report",
and instead I would need to use 'getopts' or use 'case' while
parsing the positional parameters. And accordingly rework the
'coccicheck' shell scripts at many places.

Anyway, let me know the final decision of what should be done ?
Make ? CMake ? sanitycheck  ?

CC @carlescufi

> If there are no plans for making this portable, and it does not use any information from the build system I don't see any benefits of Make over CMake, but several downsides; e.g. an inconvenient CLI.

The only info that we use is $ZEPHYR_BASE from build system
environment to determine the root source code directory!

Portability only on *nix and its derivatives + macOS.
Windows impossible as of now...

And now I wonder how difficult would it be to set up git, build-essentials,
kernel development environment, ... on Windows using Powershell.

Running Eclipse, Code Composer Studio IDEs is different story though...

> As an aside, I was not able to make it run using coccinelle 1.0.4.deb-2. It prints a bunch of compare pointer with 0 warnings and then errors out with "coccicheck failed".

As @JuliaLawall already pointed, its old!

himanshu@himanshu-Vostro-3559:~/coccinelle$ git diff --shortstat 1.0.4 master
 829 files changed, 67399 insertions(+), 18373 deletions(-)

And it is strongly encouraged to use the latest source beacause many
times it is usually that some package supplied by distribution is old or
missing([1]because the user didn't installed it correctly). It just wastes
a lot of time on cocci-mailing list wondering where could be the
problem. 

Also, we also facility in Coccinelle to debug issues as
mentioned in cocccinelle.rst:

https://www.kernel.org/doc/html/latest/dev-tools/coccinelle.html#debugging-coccinelle-smpl-patches

[1] https://systeme.lip6.fr/pipermail/cocci/2017-December/004799.html

-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

**Comment by ceolin** [comment]
@himanshujha199640 about cmake, ./cmake/host-tools.cmake can be a good start pointing to you. There is declared some other tools we use, we can grep to see how they are used.

File: scripts/coccinelle/null/badzero.cocci
Comment (ceolin) [inline scripts/coccinelle/null/badzero.cocci]: can you change it to always do `x == NULL` ? We're going to follow MISRA-C guidelines and this is the recommended way to check pointers.

**Comment by SebastianBoe** [comment]
> Hmm. But we don't have such facility in shell-scripts to assign
> variables via commandline "M=./path/" or "V=1" or "MODE=report",
> and instead I would need to use 'getopts' or use 'case' while
> parsing the positional parameters. And accordingly rework the
> 'coccicheck' shell scripts at many places.

That is correct.

**Review by ulfalizer — COMMENTED** [review]
Dismissing review.

FWIW, I think it's fine to keep some references to Kbuild if it makes it easier to sync with the kernel (is coccinelle maintained within it?). Need some way to integrate it with Zephyr though.

Looking at some other stuff atm.

**Review by ulfalizer — APPROVED** [review]

**Comment by himanshujha199640** [comment]
On Wed, Sep 05, 2018 at 09:51:11AM -0700, Ulf Magnusson wrote:
> ulfalizer commented on this pull request.
> 
> Dismissing review.
> 
> FWIW, I think it's fine to keep some references to kbuild if it makes it easier to sync with the kernel (is coccinelle maintained within it?). Need some way to integrate it with Zephyr though.

What kernel are you referring ? Mainline or Zephyr ?

Coccinelle is maintained at https://github.com/coccinelle/coccinelle

And coccinelle front-end `scripts/coccicheck` is maintained in the
mainline with cocci scripts present at 'scripts/coccinelle/`

I want to re-iterate that the only info we use from zephyr build env
is the $ZEPHYR_BASE to determine the root source code directory.

If you want to make it only as a "sanitychecker" like checkpatch, and
run coccinelle something like:

`himanshu@himanshu-Vostro-3559:~/zephyr$ ./scripts/coccicheck --mode
report --verbose --j3 --dir drivers/i2c/ --cocci scripts/coccinelle/null/badzero.cocci`

Then I don't see any see benefit in such a use-case since `spatch`
engine used to do the transformations in Coccinelle already has such
facilities and we could simply run instead:

`himanshu@himanshu-Vostro-3559:~/zephyr$ spatch -D report --dir drivers/i2c/ --sp-file scripts/coccinelle/null/badzero.cocci --jobs 3 --chunksize 1`

@JuliaLawall comments ?

I clueless in which direction should I proceed ?
make ? CMake ? sanitychecker ?

And for each decision I have clearly stated the pros/cons, so
I will wait a while for final decision to come in and then work
in the specified direction.

And lastly, please if possible install coccinelle from Github sources
and try it yourself to get a better idea of:

* different modes
* running only on specified directory
* parallelism
* verbose/non-verbose output

Some examples: https://github.com/coccinelle/coccinelle/tree/master/demos

$ man spatch

<snip>

NAME
       spatch - apply a semantic patch file to a set of C files

SYNOPSIS
       spatch --sp-file <SP> <files> [-o <outfile> ] [--iso-file <iso> ] [ options ]

DESCRIPTION
       spatch  is a program matching and transformation tool for C.  The programmer describes the code to match and the transforma‐
       tion to perform as a semantic patch, which looks like a standard patch, but can transform multiple files at  any  number  of
       code sites.

       Further information about spatch is available at http://coccinelle.lip6.fr/.

       Semantic  patch examples can be found at http://coccinellery.org/, and at the scripts/coccinelle directory of the Linux Ker‐
       nel source code.

</snip>

-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

**Comment by carlescufi** [comment]
>I clueless in which direction should I proceed ?
make ? CMake ? sanitychecker ?

I don't know enough about Coccinelle, but first I'd like to clarify a couple of things:

1) Do we actually need a scripts in `<zephyr>/scripts` or can the scripts in the Coccinelle repo be used instead on the Zephyr codebase directly with a vanilla `coccinelle` clone?

2) Even if we don't need the script, it might be useful to have a `<zephyr>/scripts/coccicheck` that assumes the presence of Coccinelle in the `PATH`? Or is this really not required?

**Comment by himanshujha199640** [comment]
Coccinelle as a **sanitychecker** has been added in this revision.

Changelog:

Following review request are addressed:
@dbkinder
- Moved coccinelle.rst to `doc/application/coccinelle.rst` and linked to TOC.
- Zephyr specific examples in  the doc.
- Followed Zephyr guideline of documentation.

@ulfalizer & @SebastianBoe
- No kbuild reference is used.
- Removed `make` build system and now can be used as sanity checker using `./scripts/coccicheck`

Sample tests:
https://gist.github.com/himanshujha199640/d43566b45909b20822cee3daa3e0ce9b

@ceolin 
The scripts you have written needs to have virtual rules to run various modes. Also, since
you have only added patch case {-/+} in all those scripts at `scripts/coccinelle`, consider adding
virtual rule, so that we can use those mode using `cocicheck`. Look at `array_size.cocci` that I just pushed for refernce and also look at `coccinelle.rst` for various guidelines when proposing a new semantic patch.

I assume you run your cocci scripts now using:
`$ spatch --sp-file unsigned_shift.cocci --very-quiet drivers/`

Also, we will now categorise rules into appropriate subdirectories as done here:
https://github.com/torvalds/linux/tree/master/scripts/coccinelle

Enforcing API usage like array_size.cocci would go into `api` directory.
Locking related rules into `locks` 
... and likewise for others.

Once this pull request gets merged I will convert all the scripts present at `scripts/coccinelle/` to add
various virtual rules and conform to the `coccinelle.rst` guideline of proposing a new semantic patch.

**Comment by nashif** [comment]
Can you please address the documentation issues?

**Comment by himanshujha199640** [comment]
On Sun, Sep 30, 2018 at 02:13:06PM -0700, Anas Nashif wrote:
> Can you please address the documentation issues?

What documentation issues ?

Is it about some issue failing while building docs or some
section that needs to be rephrased ?


-- 
Himanshu Jha
Undergraduate Student
Department of Electronics & Communication
Guru Tegh Bahadur Institute of Technology

**Comment by nashif** [comment]
> What documentation issues ?

https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/23417/5/tests

**Review by SebastianBoe — APPROVED** [review]

File: doc/application/coccinelle.rst
Comment (dbkinder) [inline doc/application/coccinelle.rst]: Check for consistent use of code block role shorthand for options. Sometimes short commands are using the double-backtick syntax and other times you're not.  For example, this line should say 

    set ``--verbose=1`` option

File: doc/application/coccinelle.rst
Comment (dbkinder) [inline doc/application/coccinelle.rst]: similarly, this line should say:

    the ``--jobs=<number>`` option.

File: doc/application/coccinelle.rst
Comment (dbkinder) [inline doc/application/coccinelle.rst]: you sometimes use coccicheck with and without the double-backticks

File: doc/application/coccinelle.rst
Comment (dbkinder) [inline doc/application/coccinelle.rst]: here you're using a path for the script name, where previously you just used the script name.

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by himanshujha199640** [comment]
@nashif & @dbkinder 
Documentation issues has been fixed!

File: doc/application/coccinelle.rst
Comment (dbkinder) [inline doc/application/coccinelle.rst]: Sorry, missed one. Needs to be a blank line before the body of this hlist directive.  Otherwise LGTM.

**Review by dbkinder — DISMISSED** [review]
one (last) doc issue

**Review by nashif — APPROVED** [review]

**Pull request merged**
